### PR TITLE
ENG-19900 : Additional collector metrics for ingest error

### PR DESCRIPTION
**Pull request by rakhimundhada15**

### CODE REVIEW

File: paws_collector.js [hunk 1]
```diff
@@ -479,7 +491,7 @@ class PawsCollector extends AlAwsCollector {
             ],
             Namespace: 'PawsCollectors'
         };
-        let errorCode = typeof (error) === 'object' && error.errorCode ? error.errorCode : 'unknown';
+        let errorCode = typeof (error) === 'object' && error.errorCode ? error.errorCode : `unknown_${this.collector_id}`;
```
Comment (JDragovichAlertLogic) [inline paws_collector.js:494]: @kkuzmin what do you think of adding the collector id to the "unknown" error tag? Are we going to get too many tags? I think it should be ok.

File: paws_collector.js [hunk 2]
```diff
@@ -450,6 +454,14 @@ class PawsCollector extends AlAwsCollector {
         this.reportDDMetric('api_throttling', 1)
         return cloudwatch.putMetricData(params, callback);
     };
+    /**
+     * Report the error to Ingest api service and show case on DDMetrics
+     * @param error 
+     */
+    reportErrorToIngestApi(error) {
```
Comment (JDragovichAlertLogic) [inline paws_collector.js:461]: Not sure how this is reporting the error to ingest. Are we handling this through a collector status? 

Alos shouldn't this have some sort of callback? arent we going to have a race condition with the async callback?

like this https://github.com/alertlogic/paws-collector/blob/6a6dcc561aa2bee5fa9085ed4ab3016637ba8c89/paws_collector.js#L387-L389

File: paws_collector.js [hunk 3]
```diff
@@ -392,8 +392,12 @@ class PawsCollector extends AlAwsCollector {
             },
             function(logs, privCollectorState, nextInvocationTimeout, asyncCallback) {
                 console.info('PAWS000200 Log events received ', logs.length);
-                return collector.processLog(logs, collector.pawsFormatLog.bind(collector), null, function(err) {
-                    return asyncCallback(err, privCollectorState, nextInvocationTimeout);
+                return collector.processLog(logs, collector.pawsFormatLog.bind(collector), null, (err) => {
+                    if (err) {
+                        collector.reportErrorToIngestApi(err);
```
Comment (JDragovichAlertLogic) [inline paws_collector.js:397]: see comment below. I think we need to add a callback to this function

**Review by JDragovichAlertLogic — CHANGES_REQUESTED** [review]
Thanks for this Rakhi. a few comments

File: paws_collector.js (hunk 1 above)
Comment (daniel-greening) [reply paws_collector.js:494]: I've been advised against doing this in the past. Typically you want to keep instance specific dimensions out of datadog metrics, apparently due to cost issues, though maybe this has changed. 

I'll reach out to cloud ops team about it.

Comment (kkuzmin) [reply paws_collector.js:494]: I haven't heard anything changed about tags. I advise to not do this.
If required we should always be able to find collector ID in service logs in kibana

Comment (JDragovichAlertLogic) [reply paws_collector.js:494]: Sounds reasonable.

The one issue with these client errors is that they never reach our back end. So if we see client errors in DD, we don't really have a good way of finding out which collector is causing the error.

crazy idea. could we put all the cloudwatch logs for paws collectors in search?

**Review by JDragovichAlertLogic — CHANGES_REQUESTED** [review]

**Review by daniel-greening — APPROVED** [review]

**Review by daniel-greening — APPROVED** [review]

**Pull request merged**